src/Test/Utilities/Portable/Metadata/ILValidation.cs

```diff
@@ -3,16 +3,182 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.IO;
 using System.Linq;
+using System.Reflection;
 using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Security.Cryptography;
 using System.Text;
+using System.Threading;
 using System.Xml;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis;
 using Microsoft.Metadata.Tools;
 
 namespace Roslyn.Test.Utilities
 {
     public static class ILValidation
     {
+        /// <summary>
+        /// Validates that the given stream is marked as signed, the signature matches
+        /// the public key, and the header checksum is correct.
+        /// </summary>
+        public static bool IsStreamFullSigned(Stream moduleContents)
+        {
+            moduleContents.Position = 0;
+
+            using (var metadata = ModuleMetadata.CreateFromStream(moduleContents))
+            {
+                var metadataReader = metadata.MetadataReader;
+                var peReader = metadata.Module.PEReaderOpt;
+                var peHeaders = peReader.PEHeaders;
+                var flags = peHeaders.CorHeader.Flags;
+
+                bool is32bit = peHeaders.PEHeader.Magic == PEMagic.PE32;
+                const int SectionHeaderSize = 40;
+
+                if (CorFlags.StrongNameSigned != (flags & CorFlags.StrongNameSigned))
+                {
+                    return false;
+                }
+
+                var snDirectory = peReader.PEHeaders.CorHeader.StrongNameSignatureDirectory;
+                int rva = snDirectory.RelativeVirtualAddress;
+                int size = snDirectory.Size;
+                ImmutableArray<byte> signature = peReader.GetSectionData(rva).GetContent(0, size);
+                if (!peHeaders.TryGetDirectoryOffset(snDirectory, out int snOffset))
+                {
+                    return false;
+                }
+
+                moduleContents.Position = 0;
+                int peSize = checked((int)moduleContents.Length);
```

Comment:
This should be wrapped in a `try / catch / return false`. If the stream signature isn't valid I would expect false to be returned here, not have an overflowexception thrown.